core/src/main/scala/cats/ErrorControl.scala

```diff
@@ -0,0 +1,98 @@
+package cats
+
+import cats.data._
+
+trait ErrorControl[F[_], G[_], E] extends Serializable {
+  def monadErrorF: MonadError[F, E]
+  def monadG: Monad[G]
+
+  def controlError[A](fa: F[A])(f: E => G[A]): G[A]
+
+  def accept[A](ga: G[A]): F[A]
+
+  def control[A, B](fa: F[A])(f: Either[E, A] => G[B]): G[B] =
+    monadG.flatMap(trial(fa))(f)
+
+  def trial[A](fa: F[A]): G[Either[E, A]] =
+    intercept(monadErrorF.map(fa)(Right(_): Either[E, A]))(Left(_))
+
+  def trialT[A](fa: F[A]): EitherT[G, E, A] =
+    EitherT(trial(fa))
+
+  def intercept[A](fa: F[A])(f: E => A): G[A] =
+    controlError(fa)(f andThen monadG.pure)
+
+  def absolve[A](gea: G[Either[E, A]]): F[A] =
+    monadErrorF.flatMap(accept(gea))(_.fold(monadErrorF.raiseError, monadErrorF.pure))
+
+  def assure[A](ga: G[A])(error: => E)(predicate: A => Boolean): F[A] =
+    assureOr(ga)(_ => error)(predicate)
+
+  def assureOr[A](ga: G[A])(error: A => E)(predicate: A => Boolean): F[A] =
```

Comment:
seems to me `A => Option[E]` rather than two functions is a nicer way to go. We also only have one invocation. using two functions in one call is usually not super ergonomic in scala.

Reply:
I assume you mean `Option[A] => E`? I think we could just get rid of what's currently `assure`, as `assureOr` covers all the use cases. I just kept both for the sake of symmetry in `ensure` and `ensureOr`. I'd rather just have a single function `assure` that takes `A => E` like in the blog post.